File: best-time-to-buy-iii.java
public class Solution {
    public int maxProfit(int[] prices) {
        if (prices.length == 0) return 0;
        int[] bestBefore = new int[prices.length];
        int minPrice = prices[0];
        bestBefore[0] = 0;
        for (int i = 1; i < prices.length; i++) {
            bestBefore[i] = Math.max(bestBefore[i - 1], prices[i] - minPrice);
            minPrice = Math.min(minPrice, prices[i]);
        }
        int[] bestAfter = new int[prices.length];
        int maxPrice = prices[prices.length - 1];
        for (int i = prices.length - 2; i >= 0; i--) {
            bestAfter[i] = Math.max(bestAfter[i + 1], maxPrice - prices[i]);
            maxPrice = Math.max(maxPrice, prices[i]);
        }
        int ans = 0;
        for (int i = 0; i < prices.length; i++) {
            if (bestBefore[i] + bestAfter[i] > ans)
                ans = bestBefore[i] + bestAfter[i];
        }
        return ans;
    }
}

File: best-time-to-buy-iii.cc
#include <vector>
using namespace std;

class Solution {
public:
  int maxProfit(vector<int> &prices) {
    if (prices.empty()) return 0;
    vector<int> front_profit(prices.size());
    vector<int> back_profit(prices.size());
    int min_price = prices[0];
    for (int i = 1; i < prices.size(); i++) {
      front_profit[i] = max(front_profit[i - 1], prices[i] - min_price);
      min_price = min(prices[i], min_price);
    }
    int max_price = prices.back();
    for (int i = prices.size() - 2; i >= 0; i--) {
      back_profit[i] = max(back_profit[i + 1], max_price - prices[i]);
      max_price = max(prices[i], max_price);
    }
    int ans = back_profit[0];  // One transaction.
    for (int i = 1; i < prices.size(); i++) {
      ans = max(ans, front_profit[i - 1] + back_profit[i]);
    }
    return ans;
  }
};
